### PR TITLE
IGNITE-15814 [Native Persistence 3.0] Data Structures porting

### PR DESCRIPTION
**Pull request by tkalkirill**
https://issues.apache.org/jira/browse/IGNITE-15814

### CODE REVIEW

File: modules/page-memory/src/main/java/org/apache/ignite/internal/pagememory/io/PageIo.java [hunk 1]
```diff
@@ -54,86 +55,75 @@
...
-    /**
-     * Total size of common header, including reserved bytes.
-     */
+    /** Total size of common header, including reserved bytes. */
     public static final int COMMON_HEADER_END = RESERVED_3_OFF + Long.BYTES;
 
-    /**
-     * IO version.
-     */
+    /* All the page types. */
```
Comment (ibessonov) [inline modules/page-memory/src/main/java/org/apache/ignite/internal/pagememory/io/PageIo.java:97]: All references to specific page types should be removed from this class

File: modules/page-memory/src/main/java/org/apache/ignite/internal/pagememory/io/PageIo.java [hunk 2]
```diff
@@ -54,86 +55,75 @@
...
+    public static final short T_META = 11;
+
+    /** Type page list meta. */
+    public static final short T_PAGE_LIST_META = 12;
+
+    /** Type page list node. */
+    public static final short T_PAGE_LIST_NODE = 13;
+
+    /** Simple (just byte array) data page type. */
+    public static final short T_DATA_PART = 32;
```
Comment (ibessonov) [inline modules/page-memory/src/main/java/org/apache/ignite/internal/pagememory/io/PageIo.java:112]: Also, please remove unused page types.

File: modules/page-memory/src/test/java/org/apache/ignite/internal/pagememory/freelist/TestDataPageIo.java [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+ */
+class TestDataPageIo extends AbstractDataPageIo<TestDataRow> {
+    /** I/O versions. */
+    static final IoVersions<TestDataPageIo> VERSIONS = new IoVersions<>(new TestDataPageIo());
+
+    /**
+     * Private constructor.
+     */
+    private TestDataPageIo() {
+        super(T_DATA, 1);
```
Comment (ibessonov) [inline modules/page-memory/src/test/java/org/apache/ignite/internal/pagememory/freelist/TestDataPageIo.java:37]: I believe that we need to have a reserved segment of types for tests specifically. Current usage of the constant from PageIo is just wrong

File: modules/page-memory/src/test/resources/META-INF/services/org.apache.ignite.internal.pagememory.io.PageIoModule [hunk 4]
```diff
@@ -15,3 +15,4 @@
 # limitations under the License.
 #
 org.apache.ignite.internal.pagememory.TestPageIoModule
+org.apache.ignite.internal.pagememory.PageMemoryIoModule
```
Comment (ibessonov) [inline modules/page-memory/src/test/resources/META-INF/services/org.apache.ignite.internal.pagememory.io.PageIoModule:18]: Wait a minute, why did you add this here?

**Review by ibessonov — CHANGES_REQUESTED** [review]

File: modules/core/src/main/java/org/apache/ignite/internal/util/ArrayUtils.java [hunk 5]
```diff
@@ -57,155 +58,137 @@
...
+
+            default:
+                return new boolean[len];
         }
     };
 
+    /**
+     * Stub.
+     */
+    private ArrayUtils() {
```
[the remark this answers is not in the document]
Comment (tkalkirill) [reply modules/core/src/main/java/org/apache/ignite/internal/util/ArrayUtils.java:183]: Deleted it.

File: modules/core/src/main/java/org/apache/ignite/internal/util/ArrayUtils.java [hunk 6]
```diff
@@ -57,155 +58,137 @@
...
+        // No op.
+    }
+
     /**
      * Returns {@code true} if {@code null} or an empty array is provided, {@code false} otherwise.
      *
      * @param arr Array to check.
      * @param <T> Array element type.
-     * @return {@code true} if {@code null} or an empty array is provided, {@code false} otherwise.
```
[the remark this answers is not in the document]
Comment (tkalkirill) [reply modules/core/src/main/java/org/apache/ignite/internal/util/ArrayUtils.java:208]: Returned back.

File: modules/core/src/main/java/org/apache/ignite/internal/util/ArrayUtils.java [hunk 7]
```diff
@@ -340,9 +316,58 @@ public static boolean nullOrEmpty(boolean[] arr) {
...
+
+        if (idx == len - 1) {
+            return Arrays.copyOfRange(arr, 0, len - 1);
+        }
+
+        if (idx == 0) {
+            return Arrays.copyOfRange(arr, 1, len);
+        }
+
+        T[] res = (T[]) Array.newInstance(arr.getClass().getComponentType(), len - 1);
```
[the remark this answers is not in the document]
Comment (tkalkirill) [reply modules/core/src/main/java/org/apache/ignite/internal/util/ArrayUtils.java:338]: Added.

File: modules/core/src/main/java/org/apache/ignite/internal/util/ArrayUtils.java [hunk 8]
```diff
@@ -340,9 +316,58 @@ public static boolean nullOrEmpty(boolean[] arr) {
...
+
+        if (idx == len - 1) {
+            return Arrays.copyOfRange(arr, 0, len - 1);
+        }
+
+        if (idx == 0) {
+            return Arrays.copyOfRange(arr, 1, len);
+        }
+
+        long[] res = new long[len - 1];
```
[the remark this answers is not in the document]
Comment (tkalkirill) [reply modules/core/src/main/java/org/apache/ignite/internal/util/ArrayUtils.java:366]: Added.

File: modules/core/src/main/java/org/apache/ignite/internal/util/IgniteLongList.java [hunk 9]
```diff
@@ -0,0 +1,567 @@
...
+import java.util.Arrays;
+import java.util.NoSuchElementException;
+import org.apache.ignite.internal.tostring.S;
+import org.apache.ignite.lang.IgniteStringBuilder;
+import org.jetbrains.annotations.Nullable;
+
+/**
+ * Minimal list API to work with primitive longs. This list exists to avoid boxing/unboxing when using standard list from Java.
+ */
+public class IgniteLongList implements Externalizable {
```
[the remark this answers is not in the document]
Comment (tkalkirill) [reply modules/core/src/main/java/org/apache/ignite/internal/util/IgniteLongList.java:35]: Added

File: modules/page-memory/src/main/java/org/apache/ignite/internal/pagememory/datastructure/DataStructure.java [hunk 10]
```diff
@@ -0,0 +1,469 @@
...
+import org.apache.ignite.internal.util.IgniteUtils;
+import org.apache.ignite.lang.IgniteInternalCheckedException;
+import org.jetbrains.annotations.Nullable;
+
+/**
+ * Base class for all the data structures based on {@link PageMemory}.
+ */
+public abstract class DataStructure {
+    /** For tests. */
+    public static Random rnd;
```
[the remark this answers is not in the document]
Comment (tkalkirill) [reply modules/page-memory/src/main/java/org/apache/ignite/internal/pagememory/datastructure/DataStructure.java:52]: Added.

File: modules/page-memory/src/main/java/org/apache/ignite/internal/pagememory/freelist/AbstractFreeList.java [hunk 11]
```diff
@@ -0,0 +1,923 @@
...
+
+        rmvRow = new RemoveRowHandler(grpId == 0);
+
+        int pageSize = pageMem.pageSize();
+
+        assert isPow2(pageSize) : "Page size must be a power of 2: " + pageSize;
+        assert isPow2(BUCKETS);
+        assert BUCKETS <= pageSize : pageSize;
+
+        // TODO this constant is used because currently we cannot reuse data pages as index pages
```
[the remark this answers is not in the document]
Comment (tkalkirill) [reply modules/page-memory/src/main/java/org/apache/ignite/internal/pagememory/freelist/AbstractFreeList.java:415]: Added

File: modules/page-memory/src/main/java/org/apache/ignite/internal/pagememory/freelist/PagesList.java [hunk 12]
```diff
@@ -0,0 +1,2172 @@
...
+        if (io.getNextId(pageAddr) != 0L) {
+            return false; // Splitted.
+        }
+
+        long nextId;
+
+        long prevId = pageId;
+        long prevAddr = pageAddr;
+
+        IgniteLongList locked = null; // TODO may be unlock right away and do not keep all these pages locked?
```
[the remark this answers is not in the document]
Comment (tkalkirill) [reply modules/page-memory/src/main/java/org/apache/ignite/internal/pagememory/freelist/PagesList.java:1080]: Added

File: modules/page-memory/src/main/java/org/apache/ignite/internal/pagememory/freelist/PagesList.java [hunk 13]
```diff
@@ -0,0 +1,2172 @@
...
+            return S.toString(SingletonReuseBag.class, this, "pageId", hexLong(pageId));
+        }
+    }
+
+    /**
+     * Class to store page-list cache onheap.
+     */
+    public static class PagesCache {
+        /** Pages cache maximum size. */
+        public static final String IGNITE_PAGES_LIST_CACHING_MAX_CACHE_SIZE = "IGNITE_PAGES_LIST_CACHING_MAX_CACHE_SIZE";
```
[the remark this answers is not in the document]
Comment (tkalkirill) [reply modules/page-memory/src/main/java/org/apache/ignite/internal/pagememory/freelist/PagesList.java:1902]: Added

File: modules/page-memory/src/main/java/org/apache/ignite/internal/pagememory/freelist/io/PagesListNodeIo.java [hunk 14]
```diff
@@ -0,0 +1,276 @@
...
+import static org.apache.ignite.internal.pagememory.util.PageUtils.putLong;
+import static org.apache.ignite.internal.pagememory.util.PageUtils.putShort;
+
+import org.apache.ignite.internal.pagememory.io.IoVersions;
+import org.apache.ignite.internal.pagememory.io.PageIo;
+import org.apache.ignite.lang.IgniteStringBuilder;
+
+/**
+ * Pages list node IO.
+ * TODO optimize: now we have slow {@link #removePage(long, long)}.
```
[the remark this answers is not in the document]
Comment (tkalkirill) [reply modules/page-memory/src/main/java/org/apache/ignite/internal/pagememory/freelist/io/PagesListNodeIo.java:33]: Added.

File: modules/page-memory/src/test/java/org/apache/ignite/internal/pagememory/PageIoRegistryTest.java [hunk 15]
```diff
@@ -44,7 +46,7 @@ void testResolve() throws Exception {
         assertThrows(IgniteInternalCheckedException.class, () -> ioRegistry.resolve(TEST_PAGE_TYPE, TEST_PAGE_VER));
 
         // Load all PageIOModule-s from the classpath.
-        ioRegistry.loadFromServiceLoader();
+        ioRegistry.loadAllFromServiceLoader();
```
[the remark this answers is not in the document]
Comment (tkalkirill) [reply modules/page-memory/src/test/java/org/apache/ignite/internal/pagememory/PageIoRegistryTest.java:49]: Returned the original name

File: modules/page-memory/src/test/resources/META-INF/services/org.apache.ignite.internal.pagememory.io.PageIoModule (hunk 4 above)
Comment (tkalkirill) [reply modules/page-memory/src/test/resources/META-INF/services/org.apache.ignite.internal.pagememory.io.PageIoModule:18]: Yes, he is superfluous here, removed.

File: modules/page-memory/src/main/java/org/apache/ignite/internal/pagememory/io/PageIoRegistry.java [hunk 16]
```diff
@@ -33,58 +33,77 @@
...
 
-                ioVersions[ios.getType()] = ios;
-            }
+    /**
+     * Loads {@link IoVersions}'s.
+     *
+     * @return {@code this} for code chaining.
+     * @throws IllegalStateException If there's an invalid page type or several {@link IoVersions} instances for the same type.
+     */
+    public PageIoRegistry load(IoVersions<?>... versions) {
```
[the remark this answers is not in the document]
Comment (tkalkirill) [reply modules/page-memory/src/main/java/org/apache/ignite/internal/pagememory/io/PageIoRegistry.java:57]: Replace with **org.apache.ignite.internal.pagememory.PageIoRegistryEx**

File: modules/page-memory/src/main/java/org/apache/ignite/internal/pagememory/io/PageIo.java (hunk 2 above)
Comment (tkalkirill) [reply modules/page-memory/src/main/java/org/apache/ignite/internal/pagememory/io/PageIo.java:108]: Removed

File: modules/page-memory/src/main/java/org/apache/ignite/internal/pagememory/io/PageIo.java (hunk 1 above)
Comment (tkalkirill) [reply modules/page-memory/src/main/java/org/apache/ignite/internal/pagememory/io/PageIo.java]: Removed

File: modules/page-memory/src/test/java/org/apache/ignite/internal/pagememory/freelist/TestDataPageIo.java (hunk 3 above)
Comment (tkalkirill) [reply modules/page-memory/src/test/java/org/apache/ignite/internal/pagememory/freelist/TestDataPageIo.java:37]: Changed

File: modules/page-memory/src/main/java/org/apache/ignite/internal/pagememory/freelist/PagesList.java [hunk 17]
```diff
@@ -866,7 +862,7 @@ protected final void put(
 
                 assert getPageId(tailAddr) == tailId
                         : "tailId = " + hexLong(tailId) + ", pageId = " + hexLong(getPageId(tailAddr));
-                assert getType(tailAddr) == T_PAGE_LIST_NODE
+                assert getType(tailAddr) == PagesListNodeIo.VERSIONS.getType()
```
Comment (ibessonov) [inline modules/page-memory/src/main/java/org/apache/ignite/internal/pagememory/freelist/PagesList.java:865]: There can be a constant for this in `PagesListNodeIo`, it's fine

File: modules/page-memory/src/main/java/org/apache/ignite/internal/pagememory/freelist/PagesList.java [hunk 18]
```diff
@@ -1379,14 +1377,7 @@ private byte getFlag(@Nullable IoVersions<?> initIoVers) {
...
 
-            switch (pageIo.getType()) {
-                case T_META:
-                case T_DATA:
-                case T_DATA_PART:
-                    return FLAG_DATA;
-                default:
-                    return defaultPageFlag;
-            }
+            return pageIo.getFlag() == FLAG_DATA ? FLAG_DATA : defaultPageFlag;
```
Comment (ibessonov) [inline modules/page-memory/src/main/java/org/apache/ignite/internal/pagememory/freelist/PagesList.java:1380]: I'm pretty sure that you can return `pageIo.getFlag()`, we don't have INDEX flag enymore

File: modules/page-memory/src/main/java/org/apache/ignite/internal/pagememory/freelist/io/PagesListMetaIo.java [hunk 19]
```diff
@@ -51,7 +52,7 @@
      * @param ver Page format version.
      */
     private PagesListMetaIo(int ver) {
-        super(T_PAGE_LIST_META, ver);
+        super(12, ver, FLAG_AUX);
```
Comment (ibessonov) [inline modules/page-memory/src/main/java/org/apache/ignite/internal/pagememory/freelist/io/PagesListMetaIo.java:55]: So yeah, please introduce a constant. Why is it 12? We could start numbering them from 1

File: modules/page-memory/src/test/java/org/apache/ignite/internal/pagememory/PageIoRegistryEx.java [hunk 20]
```diff
@@ -0,0 +1,57 @@
...
+
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+import org.apache.ignite.internal.pagememory.io.IoVersions;
+import org.apache.ignite.internal.pagememory.io.PageIoRegistry;
+
+/**
+ * Extension (for tests) of {@link PageIoRegistry} to be able to add/replace {@link IoVersions} not through {@link #loadFromServiceLoader}.
+ */
+public class PageIoRegistryEx extends PageIoRegistry {
```
Comment (ibessonov) [inline modules/page-memory/src/test/java/org/apache/ignite/internal/pagememory/PageIoRegistryEx.java:28]: Can we name it `TestPageIoRegistry`? "Ex" has nothing to do with its purpose

Comment (tkalkirill) [reply modules/page-memory/src/test/java/org/apache/ignite/internal/pagememory/PageIoRegistryEx.java:28]: Renamed

File: modules/page-memory/src/main/java/org/apache/ignite/internal/pagememory/freelist/PagesList.java (hunk 17 above)
Comment (tkalkirill) [reply modules/page-memory/src/main/java/org/apache/ignite/internal/pagememory/freelist/PagesList.java:865]: Peplace it

**Review by ibessonov — APPROVED** [review]
